File: 61. Rotate List.js
// https://leetcode.com/problems/rotate-list/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if(!head) return null
    k %= length(head);
    if(k === 0) return head;

    let [slow, fast] = [head, head];
    while(k--) fast = fast.next;

    while(fast.next){
        slow = slow.next;
        fast = fast.next;
    }

    const newHead = slow.next;
    slow.next = null;
    fast.next = head;
    return newHead;

};

const length = (head) => {
    let result = 0;
    while(head){
        result++;
        head = head.next;
    }
    return result;
}

File: 61. Rotate List.go
// https://leetcode.com/problems/rotate-list/

package leetcode

// ListNode is the standard leetcode definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}

	length, tmp := 0, head
	for tmp != nil {
		length++
		tmp = tmp.Next
	}

	k %= length
	if k == 0 {
		return head
	}

	slow, fast := head, head
	for i := 0; i < k; i++ {
		fast = fast.Next
	}

	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}

	newHead := slow.Next
	slow.Next = nil
	fast.Next = head
	return newHead
}
